returns/interfaces/failable.py

```diff
@@ -0,0 +1,230 @@
+    """
+
+    _laws: ClassVar[Sequence[Law]] = (
+        Law3(_FailableLawSpec.rescue_short_circuit_law),
+    )
+
+
+#: Type alias for kinds with two type arguments.
+Failable2 = FailableN[_FirstType, _SecondType, NoReturn]
+
+#: Type alias for kinds with three type arguments.
+Failable3 = FailableN[_FirstType, _SecondType, _ThirdType]
+
+
+@final
+class _SingleFailableLawSpec(LawSpecDef):
+    """
+    Single Failable laws.
+
+    We need to be sure that ``.map`` and ``.bind``
+    works correctly for ``empty`` property.
+    """
+
+    @law_definition
+    def map_short_circuit_law(
+        container: 'SingleFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[[_FirstType], _NewFirstType],
+    ) -> None:
+        """Ensures that you cannot map from the `empty` property."""
+        assert_equal(
+            container.empty,
+            container.empty.map(function),
+        )
+
+    @law_definition
+    def bind_short_circuit_law(
+        container: 'SingleFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[
+            [_FirstType],
+            KindN['SingleFailableN', _NewFirstType, _SecondType, _ThirdType],
+        ],
+    ) -> None:
+        """Ensures that you cannot bind from the `empty` property."""
+        assert_equal(
+            container.empty,
+            container.empty.bind(function),
+        )
+
+
+class SingleFailableN(
+    FailableN[_FirstType, _SecondType, _ThirdType],
+    Lawful['SingleFailableN[_FirstType, _SecondType, _ThirdType]'],
+):
+    """
+    Base type for types that have just only one failed value.
+
+    Like ``Maybe`` types where the only failed value is ``Nothing``.
+    """
+
+    _laws: ClassVar[Sequence[Law]] = (
+        Law2(_SingleFailableLawSpec.map_short_circuit_law),
+        Law2(_SingleFailableLawSpec.bind_short_circuit_law),
+    )
+
+    #: Alias for failed value
+    empty: ClassVar['SingleFailableN[_FirstType, _SecondType, _ThirdType]']
+
+
+#: Type alias for kinds with two types arguments.
+SingleFailable2 = FailableN[_FirstType, _SecondType, NoReturn]
+
+#: Type alias for kinds with three type arguments.
+SingleFailable3 = FailableN[_FirstType, _SecondType, _ThirdType]
+
+
+@final
+class _DiverseFailableLawSpec(LawSpecDef):
+    """
+    Diverse Failable laws.
+
+    We need to be sure that ``.map``, ``.bind`` and ``.alt``
+    works correctly for both success and failure types.
+    """
+
+    @law_definition
+    def map_short_circuit_law(
+        raw_value: _SecondType,
+        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[[_FirstType], _NewFirstType],
+    ) -> None:
+        """Ensures that you cannot map a failure."""
+        assert_equal(
+            container.from_failure(raw_value),
+            container.from_failure(raw_value).map(function),
+        )
+
+    @law_definition
+    def bind_short_circuit_law(
```

Comment:
Currently we have only `bind`, `map`, `alt`, and `rescue` covered. But, `apply` should also short circuit for `Failable` subtypes.

Let's add `.apply` as well. To both interfaces.

Reply:
Ok, I made. I got two errors with `IOResult` with apply when the law was:
```python
class Laws:
    @law_definition
    def apply_short_circuit_law(
        raw_value: _SecondType,
        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
        wraped_function: KindN[
            'DiverseFailableN',
            Callable[[_FirstType], _NewFirstType],
            _SecondType,
            _ThirdType,
        ],
    ) -> None:
        """Ensures that you cannot apply a failure."""
        assert_equal(
            container.from_failure(raw_value),
            container.from_failure(raw_value).apply(wraped_function),
        )
```

I modified to:
```python
class Laws:
    @law_definition
    def apply_short_circuit_law(
        raw_value: _SecondType,
        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
        function: Callable[[_FirstType], _NewFirstType],
    ) -> None:
        """Ensures that you cannot apply a failure."""
        assert_equal(
            container.from_failure(raw_value),
            container.from_failure(raw_value).apply(container.from_value(function)),
        )
```
And now it's working!!